thrasher-corp/gocryptotrader
# Engine QA

# Description

Several fixes/QA included in this PR, specifically:

1. Update docker Golang version to 1.13 and allow for the ability to use gctcli via docker
2. Update GitHub issue/pull request templates and fix markdown linter issues
3. Fix exchange template tool
4. Fix documentation contributor generation to include missing/non-retrievable people
5. Fix exchange wrapper coverage tool
6. Add missing exchanges to exchange_wrapper_issues config
7. Trim Slack huge struct to only fields we need
8. Make certain strings consts if used multiple times
9. Update EXCHANGE_API.md based on new order package
10. Introduce new flagset for dryrun param interaction. This means that you can specify commands like -enableallpairs or -exchangeverbose and by default the bot will induce dryrun=true. Otherwise, if you really wish for that, you can set -dryrun=false with those commands and it will disable dryrun which will overwrite your config file.
11. Link up syncer items from cmdline -> engine syncer
12. Show logging file output if enabled (just like how we show the data dir)
13. Show which exchange is sending unknown websocket messages and use consistent pair formatting
14. Rid the codebase of intermediate vars (especially for orderbook data)
15. Change all market orders to limit, or when not necessary bodge the value in case someone accidentally enables canManipulateRealOrders and runs all package tests
16. Use better structs for processing orderbook data (in Binance)
17. Made Bitstamp trading fee pairs dynamic instead of hardcoded (same with account balance)
18. Fix Bitstamp misc structs/Unmarshal errors and used built in currency types instead of inline strings
19. Add Bitstamp test time parsing test
20. Made meaningful const vars rather than "1" or "0" for OKEX/Bitstamp
21. Add BTSE time parsing test and made all funcs use it
22. Add missing err check for exchange setup calls and also in the exchange wrapper template
23. Protect Coinbene/LBank auth requests by using AllowAuthenticatedRequest()
24. Expose SupportsAsset usage for IBotInterface
25. Use Errorf instead of Warnf for time processing errors across the codebase
26. Neaten up code where necessary

# Change


# How Has This Been Tested?

go test ./... -race

# Checklist:

- [X] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool 
- [X] New and existing unit tests pass locally and on Travis with my changes
